documentation/Advanced Features/Objects and Extensions.md

```diff
@@ -0,0 +1,269 @@
+methods to an object, you can use extension methods (covered later).
+
+There is also no inheritance, meaning that composition is the only way to reuse
+code. This is not a bad thing, as it makes the implementation of objects much
+simpler.
+
+## Defining Objects
+
+To define an object:
+
+```
+#:O ObjectName | #$restrictedMember #=privateMember <data> #$restrictedMemberWithValue
+<data> #=privateMemberWithValue #!publicMember}
+```
+
+* `ObjectName` is the name of the object. This can be any valid variable name. It
+is recommended to use PascalCase for object names.
+
+Object names cannot be one of `num`, `str`, `lst` and `fun`, as these are reserved
+for the 4 built-in types.
+
+* `restrictedMember` is a member that can be read publicly, but not written to. It
+can be updated by the object itself, but not by the user. This essentially creates
+an implicit getter for the member.
+* `privateMember` is a member that can only be read and written to by the object.
+Attempting to access it from outside the object will result in an error.
+* `<data>` is any single element. It'll be used as a default value for the
+member.
+
+Those keen among us (!) will notice that the member syntax is similar to variable
+syntax. This is intentional, as it creates a consistency with existing syntax.
+Further, it creates a system of mmemonics for remembering what each member does:
+
+* `#$` was chosen as the restricted member sigil because it indicates "getting" a
+variable.
+* `#=` was chosen as the private member sigil because it indicates "setting" a
+variable, which is something that should be done in a private context.
+* `#!` was chosen as the public member sigil because it indicates danger - public
+members are typically a bad idea. It doesn't have the same meaning as `#!` for
+constants, but the `!` was appropriate to signify the danger of public members.
+
+Members without a default value will be set from the stack when the object is
+created.
+
+## Using Objects
+
+To create a new object:
+
+```
+#$ObjectName Ė
+```
+
+Defining an object creates a variable with the same name as the object. This
+pushes a constructor for the object onto the stack. The `Ė` element calls the
+constructor and pushes the created object onto the stack. This is analogous to
+calling a function, which is intentional - there's no need for a `new` keyword
+when you can just call the constructor in a similar way to a function.
+
+To access a member of an object:
+
+```
+## Assuming the object is on the stack
+"memberName" «
+```
+
+This pushes the value of the member onto the stack. If the member is a private
+member, this will fail, unless inside an extension method.
+
+To set a member of an object:
+
+```
+## Assuming the object is on the stack  
+"memberName" <value> ŀ
+```
+
+This sets the value of the member to the value on the stack. If the member is a
+restriced or private member, this will fail, unless inside an extension method.

```

Comment:
```suggestion
restricted or private member, this will fail, unless inside an extension method.

```

Maybe also say that extension methods will be covered later, and link to them